### ISSUE TITLE
Can't reproduce pytorch behavior

### ISSUE THREAD

**Issue by paholg** [LaurentMazare/tch-rs#576]
Hello! I am new to reinforcement learning, and thought I'd start with a simple example using REINFORCE / Vanilla Policy Gradient and a very simple environment.

The environment is very simple: An entity at the center of a 10 space 1d grid. It can move left or right, and is rewarded if it goes all the way right and punished if it goes all the way left. In python, it very quickly converges on always moving to the right, as can be seen here:

![2022-11-29-222032-s](https://user-images.githubusercontent.com/4908217/204722538-9c351b80-012a-4960-b1d4-4b9440a070ed.png)

[python code](https://github.com/paholg/VPG-PyTorch/blob/master/move_to_center.py)

The same algorithm, written in Rust using tch-rs, never learns:

![2022-11-29-222235-s](https://user-images.githubusercontent.com/4908217/204722858-c15335cc-96bc-4fa7-81ff-f612e4cd3034.png)

[Rust code](https://github.com/paholg/bevy_learn/blob/main/src/reinforce.rs) and [Rust env](https://github.com/paholg/bevy_learn/blob/main/examples/move_to_mid.rs).

---

As far as I can tell, the algorithms are identical. The only difference is that the Rust code uses bevy to run the loop, instead of a simple for loop. Is there some sort of global config that pytorch performs that I might be missing? Or anything else you can think of that would explain this difference?

**Comment by LaurentMazare** [comment]
Hello,
It's quite tricky to debug by just looking at the code but maybe you forgot to register the variables to the optimizer. By the way `COptimizer` is a low level api and we would recommend using the modules from `nn` instead, e.g. in your case:
```rust
let optimizer = nn::Adam::default().build(env.vs(), 1e-4).unwrap();
```
Where `vs` would be the `VarStore` from your environment that you've used to build `policy_net`. This would be used to inform the optimizer of the variables that it should perform gradient descent on.

Also you most likely already found them but there are a bunch of RL examples in this [directory](https://github.com/LaurentMazare/tch-rs/tree/main/examples/reinforcement-learning).

**Comment by paholg** [comment]
Thanks for the reply!

Switching from `COptimizer` seems to have helped some, but it's still nowhere near the level of the python version.

![2022-11-30-140734-s](https://user-images.githubusercontent.com/4908217/204918543-9e704fec-0e08-497a-a11d-01013d6edb17.png)

I did find those examples, thanks, and that's where I started, but when I couldn't get them to work, I thought I'd try something simpler. I'll try playing with those examples more, I have learned some things since then and may have been doing something wrong.

**Comment by paholg** [comment]
I just implemented the policy_gradient example for my environment, and it looks much better. I don't know enough to discern the real differences between it and my port of the python example, but I'll go ahead and close this issue. Thanks again.

![2022-11-30-145719-s](https://user-images.githubusercontent.com/4908217/204926283-13f88e4b-af19-405e-8f8b-a34558d64400.png)

**Comment by paholg** [comment]
I just implemented the policy_gradient example for my environment, and it looks much better. I don't know enough to discern the real differences between it and my port of the python example, but I'll go ahead and close this issue. Thanks again.

![2022-11-30-145719-s](https://user-images.githubusercontent.com/4908217/204926283-13f88e4b-af19-405e-8f8b-a34558d64400.png)
